src/YQCustomStatusItemSelector.cc

```diff
@@ -0,0 +1,231 @@
+/*
+  Copyright (C) 2019 SUSE LLC
+  This library is free software; you can redistribute it and/or modify
+  it under the terms of the GNU Lesser General Public License as
+  published by the Free Software Foundation; either version 2.1 of the
+  License, or (at your option) version 3.0 of the License. This library
+  is distributed in the hope that it will be useful, but WITHOUT ANY
+  WARRANTY; without even the implied warranty of MERCHANTABILITY or
+  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+  License for more details. You should have received a copy of the GNU
+  Lesser General Public License along with this library; if not, write
+  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
+  Floor, Boston, MA 02110-1301 USA
+*/
+
+
+/*-/
+
+  File:	      YQCustomStatusItemSelector.cc
+
+  Author:     Stefan Hundhammer <shundhammer@suse.de>
+
+/-*/
+
+
+#include <QToolButton>
+
+#define YUILogComponent "qt-ui"
+#include <yui/YUILog.h>
+#include <yui/YEvent.h>
+#include "utf8.h"
+#include "YQCustomStatusItemSelector.h"
+#include "YQUI.h"
+
+
+YQCustomStatusItemSelector::YQCustomStatusItemSelector( YWidget *                           parent,
+                                                        const YItemCustomStatusVector &     customStates )
+    : YQItemSelector( parent, customStates )
+{
+    preloadStatusIcons();
+}
+
+
+YQCustomStatusItemSelector::~YQCustomStatusItemSelector()
+{
+    // NOP
+}
+
+
+void YQCustomStatusItemSelector::preloadStatusIcons()
+{
+    _statusIcons.clear();
+
+    for ( int i=0; i < customStatusCount(); ++i )
```

Comment:
NP: This feels like a plain C, would it be possible to use something like `for (auto& x : foo().items()) { do_something }` ?

Reply:
No, because this uses `YItemSelector::customStatus( int index )` which acceses the custom states that this parent class stores in its PIMPL'ed _private_ struct. We get a const reference to that, so that access is the most efficient possible.

Even if the calling constructor would pass its own `YItemCustomStatusVector` to this function, that would be even worse since the parent constructor does some sanity checks and might possibly modify the data.

Just because simple things like `for` loops with int counters have been around for a while that doesn't mean that they are not useful or appropriate anymore. The new C++11 `for` loop (which should have been `foreach` anyway like in Qt) is useful when you'd have to use bulky iterators otherwise (which are typically just `container.start()` and `container.end()` anyway), but that's not the case here.

Also, IMHO `auto` variables in most cases just obfuscate the code and throw all type checks out the window; I never get to know the type of that variable, so I might make the wrong assumption what it is and do stupid things with it that involve accidential type conversion and whatnot.